tests/ErrorTest.php

```diff
@@ -253,6 +259,112 @@ public function testTwigExceptionUpdateFileAndLineTogether()
         }
     }
 
+    /**
+     * @dataProvider getErrorWithoutLineAndContextData
+     */
+    public function testErrorWithoutLineAndContext(LoaderInterface $loader, bool $debug, bool $addDebugInfo, bool $exceptionWithLineAndContext, int $errorLine)
+    {
+        $twig = new Environment($loader, ['debug' => $debug, 'cache' => false]);
+        $twig->removeCache('no_line_and_context_exception.twig');
+        $twig->removeCache('no_line_and_context_exception_include_line_5.twig');
+        $twig->removeCache('no_line_and_context_exception_include_line_1.twig');
+        $twig->addTokenParser(new class($addDebugInfo, $exceptionWithLineAndContext) extends AbstractTokenParser {
+            public function __construct(private bool $addDebugInfo, private bool $exceptionWithLineAndContext)
+            {
+            }
+
+            public function parse(Token $token)
+            {
+                $stream = $this->parser->getStream();
+                $lineno = $stream->getCurrent()->getLine();
+                $stream->expect(Token::BLOCK_END_TYPE);
+
+                return new #[YieldReady]class($lineno, $this->addDebugInfo, $this->exceptionWithLineAndContext) extends Node
+                {
+                    public function __construct(int $lineno, private bool $addDebugInfo, private bool $exceptionWithLineAndContext)
+                    {
+                        parent::__construct([], [], $lineno);
+                    }
+
+                    public function compile(Compiler $compiler): void
+                    {
+                        if ($this->addDebugInfo) {
+                            $compiler->addDebugInfo($this);
+                        }
+                        if ($this->exceptionWithLineAndContext) {
+                            $compiler
+                                ->write('throw new \Twig\Error\RuntimeError("Runtime error.", ')
+                                ->repr($this->lineno)->raw(", \$this->getSourceContext()")
+                                ->raw(");\n")
+                            ;
+                        } else {
+                            $compiler->write('throw new \Twig\Error\RuntimeError("Runtime error.");');
+                        }
+                    }
+                };
+            }
+
+            public function getTag()
+            {
+                return 'foo';
+            }
+        });
+
+        try {
+            $twig->render('no_line_and_context_exception.twig', ['line' => $errorLine]);
+            $this->fail();
+        } catch (RuntimeError $e) {
+            if (1 === $errorLine && !$addDebugInfo && !$exceptionWithLineAndContext) {
```

Comment:
Why are we reporting the exception without any line number when we don't have debug info or context and the exception is on first line of the template while we report it saying `at line 1` when the error happens on line 5 without the contextual info ? Where does this difference come from ? Is it because of the text node coming before that has some debug info on line 1 ?

Reply:
Correct, when the code is not on line 1, there is a TextNode that calls `addDebugInfo()`.

Reply:
I suggest adding a comment in the test explaining that, to avoid having future contributors wondering what could cause this (I spent quite some times reading this test before figuring out that the text node could be responsible for that)

Reply:
Done